masu/processor/providers/aws/report_downloader.py

```diff
@@ -0,0 +1,209 @@
+
+        self.logger.debug('Connecting to AWS...')
+        session = get_assume_role_session(
+            arn=self.provider.get_provider_resource_name(),
+            session='MasuDownloaderSession')
+        self.cur = session.client('cur')
+
+        defs = self.cur.describe_report_definitions()
+        self.report_name = self.provider.get_report_name()
+        report = list(filter(lambda rep: rep['ReportName'] == self.report_name,
+                             defs.get('ReportDefinitions', []))).pop()
+
+        if not report:
+            raise MasuConfigurationError('Cost and Usage Report definition not found.')
+
+        self.region = report['S3Region']
+        self.bucket = report['S3Bucket']
+        self.bucket_prefix = report['S3Prefix']
+
+        self.s3 = session.client('s3')
+
+    def _get_manifest(self, dt):
+        """
+        Download and return the CUR manifest for the given date.
+
+        Args:
+            dt (DateTime): The starting datetime object
+
+        Returns:
+            (Dict): A dict-like object serialized from JSON data.
+
+        """
+        manifest = '{}/{}-Manifest.json'.format(self._get_report_path(dt),
+                                                self.report_name)
+        manifest_file = self.download_file(manifest)
+
+        manifest_json = None
+        with open(manifest_file, 'r') as f:
+            manifest_json = json.load(f)
+
+        return manifest_json
+
+    def _get_report_path(self, dt):
+        """
+        Return path of report files.
+
+        Args:
+            dt (DateTime): The starting datetime object
+
+        Returns:
+            (String): "/prefix/report_name/YYYYMMDD-YYYYMMDD",
+                    example: "/my-prefix/my-report/19701101-19701201"
+
+        """
+        report_date_range = self._get_manifest_date_range(dt)
+        return '{}/{}/{}'.format(self.bucket_prefix,
+                                 self.report_name,
+                                 report_date_range)
+
+    def _get_manifest_date_range(self, dt):
+        """
+        Get a formatted date range string.
+
+        Args:
+            dt (DateTime): The starting datetime object
+
+        Returns:
+            (String): "YYYYMMDD-YYYYMMDD", example: "19701101-19701201"
+
+        """
+        start_month = dt.replace(day=1, second=1, microsecond=1)
+        end_month = start_month + relativedelta(months=+1)
+        timeformat = '%Y%m%d'
+        return '{}-{}'.format(start_month.strftime(timeformat),
+                              end_month.strftime(timeformat))
+
+    def download_bucket(self):
+        """
+        Bulk Download all files in an s3 bucket.
+
+        Returns:
+            (List) List of filenames downloaded.
+
+        """
+        s3_resource = boto3.resource('s3')
+        bucket = s3_resource.Bucket(self.bucket)
+        files = []
+        for s3obj in bucket.objects.all():
+            files.append(self.download_file(s3obj.key))
+        return files
```

Comment:
Downloading an entire bucket could be a lot of data. What's our use case?

Reply:
It's mostly just for development purposes. I'm uncertain if the on-boarding case is better to read-through the manifest to figure out what to download, or to simply brute-force it. It was trivial to provide both cases to allow us to experiment with it.